aptos-move/framework/aptos-framework/sources/coin.move

```diff
@@ -150,6 +187,167 @@ module aptos_framework::coin {
+            })
+        };
+        borrow_global_mut<CoinConversionMap>(@aptos_framework)
+    }
+
+    public entry fun pair_existing_fungible_asset<CoinType>(
+        coin_creator: &signer,
+        metadata: Object<Metadata>
+    ) acquires CoinConversionMap, CoinInfo {
+        // Verify the metadata is the same
+        assert!(
+            name<CoinType>() == fungible_asset::name(metadata),
+            error::invalid_argument(ECOIN_AND_FUNGIBLE_ASSET_NAME_MISMATCH)
+        );
+        assert!(
+            symbol<CoinType>() == fungible_asset::symbol(metadata),
+            error::invalid_argument(ECOIN_AND_FUNGIBLE_ASSET_SYMBOL_MISMATCH)
+        );
+        assert!(
+            decimals<CoinType>() == fungible_asset::decimals(metadata),
+            error::invalid_argument(ECOIN_AND_FUNGIBLE_ASSET_DECIMAL_MISMATCH)
+        );
+
+        // Verify both track supply or neither tracks
+        assert!(
+            option::is_some(&coin_supply<CoinType>()) == option::is_some(&fungible_asset::supply(metadata)),
+            error::invalid_argument(ESUPPLY_TRACKING_MISMATCH)
+        );
+
+        let map = borrow_conversion_map_mut();
+        let type = type_info::type_of<CoinType>();
+        let metadata_addr = object::object_address(&metadata);
+        assert!(!table::contains(&map.coin_to_fungible_asset_map, type), error::already_exists(EPAIRED_FUNGIBLE_ASSET));
+        assert!(!exists<PairedCoinType>(metadata_addr), error::already_exists(EPAIRED_COIN));
+        assert!(
+            coin_address<CoinType>() == signer::address_of(coin_creator),
+            error::permission_denied(ENOT_COIN_CREATOR)
+        );
+        assert!(
+            object::owner(metadata) == signer::address_of(coin_creator),
+            error::permission_denied(ENOT_FUNGIBLE_ASSET_METADATA_OWNER)
+        );
+        table::add(&mut map.coin_to_fungible_asset_map, type, metadata_addr);
+        move_to(&create_signer::create_signer(metadata_addr), PairedCoinType { type });
+    }
+
+    #[view]
+    /// Get the paired fungible asset metadata object of a coin type. If not exist, return option::none().
+    public fun paired_metadata<CoinType>(): Option<Object<Metadata>> acquires CoinConversionMap {
+        if (exists<CoinConversionMap>(@aptos_framework)) {
+            let map = &borrow_global<CoinConversionMap>(@aptos_framework).coin_to_fungible_asset_map;
+            let type = type_info::type_of<CoinType>();
+            if (table::contains(map, type)) {
+                option::some(object::address_to_object<Metadata>(*table::borrow(map, type)))
+            } else {
+                option::none()
+            }
+        } else {
+            option::none()
+        }
+    }
+
+    /// Get the paired fungible asset metadata object of a coin type, create if not exist.
+    public fun ensure_paired_metadata<CoinType>(): Object<Metadata> acquires CoinConversionMap, CoinInfo {
```

Comment:
nit: Rename this to start_fungible_asset_migration or something along that line? ensure_paired_metadata doesn't mean anything IMO